Remove polling from auth extensions

RMacfarlane opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #108342

bpasero commented on extensions/github-authentication/src/common/keychain.ts:
```diff
@@ -76,6 +76,11 @@ export class Keychain {
 			return Promise.resolve(undefined);
 		}
 	}
+	onDidChangePassword(listener: () => void) {
```
Not clear to me why this indirection is needed and `vscode.authentication.onDidChangePassword` cannot be used directly.

bpasero commented on extensions/github-authentication/src/github.ts:
```diff
@@ -33,52 +33,48 @@ export class GitHubAuthenticationProvider {
 		}
 
-		this.pollForChange();
+		keychain.onDidChangePassword(() => this.checkForUpdates());
```
Does this listener need to be disposed with any lifecycle?

bpasero commented on extensions/microsoft-authentication/src/AADHelper.ts:
```diff
@@ -140,7 +140,7 @@ export class AzureActiveDirectoryService {
 		}
 
-		this.pollForChange();
+		keychain.onDidChangePassword(() => this.checkForUpdates);
```
Does this listener need to be disposed with any lifecycle?

bpasero commented on extensions/microsoft-authentication/src/keychain.ts:
```diff
@@ -85,6 +85,11 @@ export class Keychain {
 			return Promise.resolve(undefined);
 		}
 	}
+	onDidChangePassword(listener: () => void) {
```
Not clear to me why this indirection is needed and `vscode.authentication.onDidChangePassword` cannot be used directly.

bpasero commented on src/vs/workbench/api/common/extHostAuthentication.ts:
```diff
@@ -24,6 +24,10 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
 	private _onDidChangeSessions = new Emitter<vscode.AuthenticationSessionsChangeEvent>();
 	readonly onDidChangeSessions: Event<vscode.AuthenticationSessionsChangeEvent> = this._onDidChangeSessions.event;
 
+
```
Extra newline

bpasero commented on src/vs/workbench/api/common/extHostAuthentication.ts:
```diff
@@ -203,4 +207,8 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
 		this._onDidChangeAuthenticationProviders.fire({ added, removed });
 		return Promise.resolve();
 	}
+	$onDidChangePassword(): Promise<void> {
```
Missing newline :)

bpasero commented on src/vs/workbench/api/common/extHostAuthentication.ts:
```diff
@@ -203,4 +207,8 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
 	}
+	$onDidChangePassword(): Promise<void> {
+		this._onDidChangePassword.fire();
+		return Promise.resolve();
```
If you use `async` on the method, you don't have to return a promise like this.

bpasero commented on src/vs/workbench/services/credentials/browser/credentialsService.ts:
```diff
@@ -6,11 +6,15 @@
 	declare readonly _serviceBrand: undefined;
 
+	private _onDidChangePassword: Emitter<void> = new Emitter();
+	onDidChangePassword = this._onDidChangePassword.event;
```
```suggestion
	readonly onDidChangePassword = this._onDidChangePassword.event;
```

bpasero commented on src/vs/workbench/services/credentials/common/credentials.ts:
```diff
@@ -17,4 +18,5 @@ export interface ICredentialsProvider {
 
 export interface ICredentialsService extends ICredentialsProvider {
 	readonly _serviceBrand: undefined;
+	onDidChangePassword: Event<void>;
```
```suggestion
	readonly onDidChangePassword: Event<void>;
```

bpasero reviewed (approved):
@RMacfarlane looks really good overall. I left some :lipstick: and lifecycle related review comments that we can do on `master`. I typically tend to let a thing `extends Disposable` whenever I deal with events or emitters to register them properly to get disposed. Even if the outer thing is never getting disposed, at least inside disposables are tracked properly. I can do this change on `master` as well next week.

bpasero commented:
@RMacfarlane after a bit more thinking, should the event possibly be more clever around this:
* only fire if the password actually changed
* carry the password around as part of the event

The reason I ask is that very likely a client of this event will ask for the new password, and that would trigger another roundtrip through all the processes until resolved. If in `setPassword` we would call `getPassword` to check if the password actually changed and add it to the event, wouldn't that make things easier?

The only downside I can think of is that `getPassword` would trigger potentially a dialog on macOS unless the user has granted this access already permanently.

RMacfarlane commented:
Thanks for reviewing! I've pushed changes addressing your feedback to master and also here.

I think we should avoid directly passing the password as part of the change event, but I agree we could do more to avoid unnecessary round trips. My concern with adding the password in the change event is that this increases its exposure - we would need to also check that the listening extension has been permitted to access this. Generally though, this event should be fired infrequently, only when a user is actually signing in or out.
